Appframework

vinceq-usgs opened the pull request:
Simplify replicate script. It is now single use and called directly from the response.php script. Crontab functionality is now removed.

emartinez-usgs commented:
- README: Installation/Configuration is not correct.
  - Do not clone directly to web infrastructure. Instead, run "dist" target of local build on a tagged version and then rsync the "dist" folder contents into the [app]/earthquake-dyfi-response directory.
  - Possibly explain what the pre-install configuration parameters mean / how one might go about answering them (do not include sensitive information here). This could also be further explained in the configuration script itself so the user is presented with useful information when installing.
  - Perhaps a link to someplace that explains ArcGIS secret keys (somewhere on ESRI site?) would be useful?
  - Pre-Install script should create all necessary output directories and set all necessary permission on these. There should be no need to check this.

- replicate.php
  - What is the use-case for the "backup" directory? Do these get purged at some point?
  - It seems most straightforward to simply build up a string containing the DYFI information to write to a file and then just write that file once to each server-specific directory.

- response.php
  - It seems you are repeatedly overwriting the `raw.input` data file (L19) on each request. This file does not get read back ever (that I see?) and may even be changed prior to a request being processed (since we may be processing many DYFI form responses at once). 
  - Suggest using template styles for alerts (L39-43). See https://earthquake.usgs.gov/theme/alert.php
  - Suggest moving ArcGIS geocoding into separate function.
  - Probably want to use `if (is_array($_POST)) { ` (L115)

vinceq-usgs commented:
Added these changes:
Fixed README installation instructions to use 'grunt dist'+rsync instead of git clone
Changed replicate.php into multiple file_put_contents
Removed ArcGIS/geocode functionality--- this should be taken care of upstream in the form
Alert message now uses alert class

emartinez-usgs commented:
As explained on the [ArcGIS terms of service](https://developers.arcgis.com/rest/geocode/api-reference/geocoding-free-vs-paid.htm), we can not remove the authenticated geocode process from response.php. My previous suggestion was to wrap that functionality up in a function and then call it in order to make response.php itself more understandable.

vinceq-usgs commented:
Eric,

1. looking at the code, it looks like the geocoding bit in response.php is
not being used.... the actual geocoding is being done upstream, in the form
itself, right?

2. How do I test response.php on dev-01? What do I change the URL value to
in the config.ini? And what URL do I point my browser to? (I assume I need
to be on the VPN to do this?)

- Vince


-- 
Vince Quitoriano (vinceq@usgs.gov)
USGS Did You Feel It?    http://earthquake.usgs.gov/earthquakes/dyfi
USGS ShakeMaps:        http://earthquake.usgs.gov/earthquakes/shakemap

On Mon, Mar 20, 2017 at 6:49 AM, Eric Martinez <notifications@github.com>
wrote:

> As explained on the ArcGIS terms of service
> <https://developers.arcgis.com/rest/geocode/api-reference/geocoding-free-vs-paid.htm>,
> we can not remove the authenticated geocode process from response.php. My
> previous suggestion was to wrap that functionality up in a function and
> then call it in order to make response.php itself more understandable.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/usgs/earthquake-dyfi-response/pull/5#issuecomment-287764213>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APXCouYXmrNeKh3VBht-jRvBW7KOeQRJks5rnoQGgaJpZM4MWAfJ>
> .
>

emartinez-usgs commented:
1. It does get used. We do perform geocoding upstream, but that is simply to display a pin on the map for the user to verify their location prior to submitting the form. Per the terms of the agreement, we must re-geocode their input location using the `forStorage=true` (and an appropriate authentication token) prior to saving this information with DYFI. We don't do authenticated requests client side because (1) it exposes our private information and (2) users may adjust their location multiple times prior to form submission and we don't want to spend too many authenticated requests on these adjustments.

If you prefer, we can remove the authenticated geocoding from response.php and strip out any geocoded coordinates, then you would be left to perform the authenticated geocode request from within the DYFI system itself. This was presented as an option back when we had to switch to ESRI but the decision was made to do the geocoding as part of response.php instead.

2. Let's chat about this offline. Send me a message when you're ready to try this out. You will need VPN to update internal configurations.

emartinez-usgs commented on README.md:
```diff
@@ -13,47 +13,40 @@ To be run on the HazDev acquisitions servers.
+3. cd to the repository root and run src/lib/pre-install to configure (this will create the target directories). It will ask for the following settings:
+    - MOUNT PATH: URL for the response.php
+    - SERVER SHORTNAME: Name of the server this is installed in
+    - WRITE DIR: Data directory. For testing, you can use [THIS_REPOSITORY]/test/ata/. For installation, you probably want [data]/earthquake-dyfi-response
```
typo: "[THIS_REPOSITORY]/test/data" ?

emartinez-usgs commented on README.md:
```diff
@@ -13,47 +13,40 @@ To be run on the HazDev acquisitions servers.
+    - MOUNT PATH: URL for the response.php
+    - SERVER SHORTNAME: Name of the server this is installed in
+    - WRITE DIR: Data directory. For testing, you can use [THIS_REPOSITORY]/test/ata/. For installation, you probably want [data]/earthquake-dyfi-response
+    - BACKEND SERVERS: Comma-delimited list of DYFI servers that will be accessing this data. The incoming entries will be copied to a separate directories each named 'incoming.[SERVER]'.
```
> ... copied to a separated directories ...

Suggest either:
1. "... copied to separate directories ..."
or
2. "... copied to a separate directory ..."

emartinez-usgs commented on src/htdocs/inc/geocode.php:
```diff
@@ -0,0 +1,32 @@
+<?php
```
This file should be added to the `lib` directory as it is a utility that is not directly requested over HTTP.

vinceq-usgs replied:
lib directory? same directory as configure.inc, pre-install etc?

emartinez-usgs commented on src/htdocs/inc/geocode.php:
```diff
@@ -0,0 +1,32 @@
+  // Geocode using ESRI ArcGIS
+  // Use authentication keys from $ini
+ 
+  function geocode($ini) {
```
Suggest parameterizing separately:
```
function geocode ($client_id, $client_secret)
```
This makes the function more readily re-usable.

emartinez-usgs commented on src/htdocs/response.php:
```diff
@@ -13,18 +13,16 @@
-$rawData = file_get_contents("php://input");
-file_put_contents("$log_dir/raw.input",$rawData);
+	$rawData = file_get_contents("php://input");
+	file_put_contents("$log_dir/raw.input",$rawData);
```
Can you explain why this log file exists? Backups of each submission are stored as well as back-end specific files. What is the purpose of `raw.input`? Be aware that multiple requests are handled concurrently and this file can not be reliably read in any meaningful way. I might suggest either:

1. Remove this `raw.input` log file.
2. Consider using the `FILE_APPEND | LOCK_EX` flags on the function call.

vinceq-usgs replied:
I've had problems where response.php is mangling the data. I need to know what the raw input is like and compare it to the output.

emartinez-usgs replied:
This approach does not give you any data integrity checks as multiple PHP threads are writing to the same file. There will be no guarantee the data has not been immediately changed.

vinceq-usgs replied:
Not guaranteed, sure, but 99% of the time the servers are quiet and I can check for problems at a glance. Also, this file gets written early in processing so it will record input if the rest of response.php is crashing consistently.

Is there a security problem with this? Or a better way to quickly check last known raw input? It doesn't sound like a resource issue.

emartinez-usgs replied:
If this is for periodic debugging/testing, it should be added as needed only on local development environments.

There is no security problem IMO.

I'm explaining it has no meaningful data. You also have backup files in the backups directory. If the script routinely crashes after receiving input, but prior to writing output, that is a problem that should be corrected.

I'm not going to fight you on this , if you want to keep it that is fine.

emartinez-usgs commented on src/htdocs/response.php:
```diff
@@ -54,148 +52,98 @@
+	$basename = implode('.',array('entry','server',$eventid,$time,$count));
+	$out_template = $data_dir . "/incoming.SERVER";
+	foreach ($backends as $backend) {
+		$dest = str_replace('SERVER',$backend,$out_template);
```
These `implode` and `str_replace` should probably just be a single `sprintf`. Seems simpler and is also faster.

emartinez-usgs commented on src/htdocs/response.php:
```diff
@@ -54,148 +52,98 @@
+$TITLE = $T['TITLE'];
+$ENCODING = 'utf-8';
+if ($form_version >= 1.2 || $windowtype == 'enabled') {
+	$TEMPLATE = "minimal";
```
This is not how you specify minimal template anymore. See: https://earthquake.usgs.gov/theme/minimal.php

emartinez-usgs commented on src/htdocs/response.php:
```diff
@@ -54,148 +52,98 @@
+}
+include_once($translate_file);
+
+$TITLE = $T['TITLE'];
```
I don't think `$TITLE` or `$ENCODING` have any effect at this point. The template has already been included (Line 25, above)

vinceq-usgs commented:
1. removing 'Server:' from output (now in output file)
2. No idea, I didn't write any of the HTML, I will try removing it.
